## Module 1 Homework (2025 cohort)

In this homework, we’re going to download finance data from various
sources and make simple calculations or analysis.

### Question 1: \[Index\] S&P 500 Stocks Added to the Index

**Which year had the highest number of additions?**

Using the list of S&P 500 companies from Wikipedia’s [S&P 500 companies
page](https://en.wikipedia.org/wiki/List_of_S%26P_500_companies),
download the data including the year each company was added to the
index.

Hint: you can use
[pandas.read_html](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html)
to scrape the data into a DataFrame.

Steps: 
1. Create a DataFrame with company tickers, names, and the year
they were added. 
2. Extract the year from the addition date and
calculate the number of stocks added each year.
3. Which year had the
highest number of additions (1957 doesn’t count, as it was the year when
the S&P 500 index was founded)?
**Write down this year as your answer (the
most recent one, if you have several records)**.

*Context*: \> “Following the announcement, all four new entrants saw
their stock prices rise in extended trading on Friday” - recent examples
of S&P 500 additions include DASH, WSM, EXE, TKO in 2025 ([Nasdaq
article](https://www.nasdaq.com/articles/sp-500-reshuffle-dash-tko-expe-wsm-join-worth-buying)).

*Additional*: How many current S&P 500 stocks have been in the index for
more than 20 years? 

When stocks are added to the S&P 500, they usually
experience a price bump as investors and index funds buy shares
following the announcement.



In [2]:
import pandas as pd


In [3]:
comp_df = pd.read_html("https://en.wikipedia.org/wiki/List_of_S%26P_500_companies")[0]
comp_df['Date added']= pd.to_datetime(comp_df['Date added'])
comp_df['year_added'] = comp_df['Date added'].dt.year
comp_df

Symbol             Security             GICS Sector  \
0      MMM                   3M             Industrials   
1      AOS          A. O. Smith             Industrials   
2      ABT  Abbott Laboratories             Health Care   
3     ABBV               AbbVie             Health Care   
4      ACN            Accenture  Information Technology   
..     ...                  ...                     ...   
498    XYL           Xylem Inc.             Industrials   
499    YUM          Yum! Brands  Consumer Discretionary   
500   ZBRA   Zebra Technologies  Information Technology   
501    ZBH        Zimmer Biomet             Health Care   
502    ZTS               Zoetis             Health Care   

                                GICS Sub-Industry    Headquarters Location  \
0                        Industrial Conglomerates    Saint Paul, Minnesota   
1                               Building Products     Milwaukee, Wisconsin   
2                           Health Care Equipment  North Chicago, Illinois   
3                                   Biotechnology  North Chicago, Illinois   
4                  IT Consulting & Other Services          Dublin, Ireland   
..                                            ...                      ...   
498  Industrial Machinery & Supplies & Components   White Plains, New York   
499                                   Restaurants     Louisville, Kentucky   
500            Electronic Equipment & Instruments   Lincolnshire, Illinois   
501                         Health Care Equipment          Warsaw, Indiana   
502                               Pharmaceuticals   Parsippany, New Jersey   

    Date added      CIK      Founded  year_added  
0   1957-03-04    66740         1902        1957  
1   2017-07-26    91142         1916        2017  
2   1957-03-04     1800         1888        1957  
3   2012-12-31  1551152  2013 (1888)        2012  
4   2011-07-06  1467373         1989        2011  
..         ...      ...          ...         ...  
498 2011-11-01  1524472         2011        2011  
499 1997-10-06  1041061         1997        1997  
500 2019-12-23   877212         1969        2019  
501 2001-08-07  1136869         1927        2001  
502 2013-06-21  1555280         1952        2013  

[503 rows x 9 columns]

In [4]:
comp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Symbol                 503 non-null    object        
 1   Security               503 non-null    object        
 2   GICS Sector            503 non-null    object        
 3   GICS Sub-Industry      503 non-null    object        
 4   Headquarters Location  503 non-null    object        
 5   Date added             503 non-null    datetime64[ns]
 6   CIK                    503 non-null    int64         
 7   Founded                503 non-null    object        
 8   year_added             503 non-null    int32         
dtypes: datetime64[ns](1), int32(1), int64(1), object(6)
memory usage: 33.5+ KB


In [5]:
yearly_additions = comp_df[comp_df['year_added'] != 1957].groupby('year_added').size()
max_additions_count = yearly_additions.max()
max_additions_year = yearly_additions[yearly_additions == max_additions_count]
max_additions_year

year_added
2016    23
2017    23
dtype: int64

### Question 2. \[Macro\] Indexes YTD (as of 1 May 2025)

**How many indexes (out of 10) have better year-to-date returns than the
US (S&P 500) as of May 1, 2025?**

Using Yahoo Finance World Indices data, compare the year-to-date (YTD)
performance (1 January-1 May 2025) of major stock market indexes for the
following countries: 
* United States - S&P 500 (^GSPC)
* China -Shanghai Composite (000001.SS) 
* Hong Kong - HANG SENG INDEX (^HSI)  
* Australia - S&P/ASX 200 (^AXJO)
* India - Nifty 50 (^NSEI)
* Canada - S&P/TSX Composite (^GSPTSE)
* Germany - DAX (^GDAXI)
* United Kingdom - FTSE 100 (^FTSE)
* Japan - Nikkei 225 (^N225)
* Mexico - IPC Mexico (^MXX)
* Brazil - Ibovespa (^BVSP)

*Hint*: use start_date=‘2025-01-01’ and end_date=‘2025-05-01’ when
downloading daily data in yfinance

Context: \> [Global Valuations: Who’s Cheap, Who’s
Not?](https://simplywall.st/article/beyond-the-us-global-markets-after-yet-another-tariff-update)
article suggests “Other regions may be growing faster than the US and
you need to diversify.”

Reference: Yahoo Finance World Indices -
https://finance.yahoo.com/world-indices/

*Additional*: How many of these indexes have better returns than the S&P
500 over 3, 5, and 10 year periods? Do you see the same trend? Note: For
simplicity, ignore currency conversion effects.)


In [6]:
import yfinance as yf
import pandas as pd

In [ ]:
# Define the indices and their tickers
indices = {
    "United States - S&P 500": "^GSPC",
    "China - Shanghai Composite": "000001.SS",
    "Hong Kong - HANG SENG INDEX": "^HSI",
    "Australia - S&P/ASX 200": "^AXJO",
    "India - Nifty 50": "^NSEI",
    "Canada - S&P/TSX Composite": "^GSPTSE",
    "Germany - DAX": "^GDAXI",
    "United Kingdom - FTSE 100": "^FTSE",
    "Japan - Nikkei 225": "^N225",
    "Mexico - IPC Mexico": "^MXX",
    "Brazil - Ibovespa": "^BVSP",
}

# Define the date range
start_date = "2025-01-01"
end_date = "2025-05-01"

# Download data for all indices
data = {name: yf.download(ticker, start=start_date, end=end_date) for name, ticker in indices.items()}

# Calculate YTD returns
returns = {}
for name, df in data.items():
    if not df.empty:
        start_price = df['Close'].iloc[0]
        end_price = df['Close'].iloc[-1]
        ytd_return = ((end_price - start_price) / start_price) * 100
        returns[name] = ytd_return.values[0]

# Convert returns to a DataFrame
returns_df = pd.DataFrame(list(returns.items()), columns=["Index", "YTD Return"])
returns_df.sort_values(by="YTD Return", ascending=False, inplace=True)

# Find how many indices have better returns than the S&P 500
sp500_return = returns["United States - S&P 500"]
better_than_sp500 = returns_df[returns_df["YTD Return"] > sp500_return]

# Output results
print("YTD Returns (1 Jan 2025 - 1 May 2025):")
print(returns_df)
print(f"Number of indices with better YTD returns than the S&P 500: {len(better_than_sp500)}")
print("Indices with better YTD returns than the S&P 500:")
print(better_than_sp500)

[*********************100%***********************]  1 of 1 completed


[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed

YTD Returns (1 Jan 2025 - 1 May 2025):
                          Index  YTD Return
9           Mexico - IPC Mexico   13.049444
2   Hong Kong - HANG SENG INDEX   12.720018
10            Brazil - Ibovespa   12.438710
6                 Germany - DAX   12.346378
7     United Kingdom - FTSE 100    2.842590
4              India - Nifty 50    2.490424
1    China - Shanghai Composite    0.504817
5    Canada - S&P/TSX Composite   -0.226126
3       Australia - S&P/ASX 200   -0.914500
0       United States - S&P 500   -5.103301
8            Japan - Nikkei 225   -8.297931

Number of indices with better YTD returns than the S&P 500: 9

Indices with better YTD returns than the S&P 500:
                          Index  YTD Return
9           Mexico - IPC Mexico   13.049444
2   Hong Kong - HANG SENG INDEX   12.720018
10            Brazil - Ibovespa   12.438710
6                 Germany - DAX   12.346378
7     United Kingdom - FTSE 100    2.842590
4              India - Nifty 50    2.490424
1    China 

### Question 3. \[Index\] S&P 500 Market Corrections Analysis

**Calculate the median duration (in days) of significant market
corrections in the S&P 500 index.**

For this task, define a correction as an event when a stock index goes
down by **more than 5%** from the closest all-time high maximum.

Steps: 
1. Download S&P 500 historical data (1950-present) using yfinance
2. Identify all-time high points (where price exceeds all previous
prices) 
3. For each pair of consecutive all-time highs, find the minimum
price in between 
4. Calculate drawdown percentages: (high - low) / high
× 100 
5. Filter for corrections with at least 5% drawdown 
6. Calculate the duration in days for each correction period 
7. Determine the 25th, 50th (median), and 75th percentiles for correction durations

*Context:* \> \* Investors often wonder about the typical length of
market corrections when deciding “when to buy the dip” ([Reddit
discussion](https://www.reddit.com/r/investing/comments/1jrqnte/when_are_you_buying_the_dip/?rdt=64135)).

> -   [A Wealth of Common Sense - How Often Should You Expect a Stock
>     Market
>     Correction?](https://awealthofcommonsense.com/2022/01/how-often-should-you-expect-a-stock-market-correction/)

*Hint (use this data to compare with your results)*: Here is the list of
top 10 largest corrections by drawdown: 
* 2007-10-09 to 2009-03-09: 56.8% drawdown over 517 days 
* 2000-03-24 to 2002-10-09: 49.1% drawdown over 929 days 
* 1973-01-11 to 1974-10-03: 48.2% drawdown over 630 days
* 1968-11-29 to 1970-05-26: 36.1% drawdown over 543 days 
* 2020-02-19 to 2020-03-23: 33.9% drawdown over 33 days 
* 1987-08-25 to 1987-12-04: 33.5% drawdown over 101 days 
* 1961-12-12 to 1962-06-26: 28.0% drawdown over 196 days 
* 1980-11-28 to 1982-08-12: 27.1% drawdown over 622 days
* 2022-01-03 to 2022-10-12: 25.4% drawdown over 282 days 
* 1966-02-09 to 1966-10-07: 22.2% drawdown over 240 days


In [1]:
import yfinance as yf
import pandas as pd
import numpy as np

In [ ]:
gspc_df = yf.download("^GSPC", start="1950-01-01", auto_adjust=True)["Close"]
gspc_df.columns = ['Close']

gspc_df['AllTimeHigh'] = gspc_df['Close'].cummax()
all_time_highs = gspc_df[gspc_df['Close'] == gspc_df['AllTimeHigh']]

In [5]:
corrections = []
all_time_highs_list = all_time_highs.index.to_list()

for i in range(len(all_time_highs_list) - 1):
    start_date = all_time_highs_list[i]
    end_date = all_time_highs_list[i + 1]
    
    segment = gspc_df.loc[start_date:end_date]
    if segment.empty:
        continue
    min_date= segment['Close'].idxmin()
    min_price = segment.loc[min_date]['Close']
    
    peak_price = gspc_df.loc[start_date]['Close']
    drawdown_pct = ((peak_price - min_price) / peak_price) * 100
    
    if drawdown_pct > 5:
        duration = (min_date - start_date).days
        corrections.append({
            'Start': start_date,
            'Trough': min_date,
            'End': end_date,
            'Peak Price': peak_price,
            'Trough Price': min_price,
            'Drawdown (%)': drawdown_pct,
            'Duration (days)': duration
        })

In [6]:
corrections_df = pd.DataFrame(corrections)
percentiles = corrections_df['Duration (days)'].quantile([0.25, 0.5, 0.75])
median_duration = percentiles[0.5]

print("Correction Duration Percentiles (in days):")
percentiles

Correction Duration Percentiles (in days):


0.25    21.5
0.50    39.0
0.75    89.0
Name: Duration (days), dtype: float64

### Question 4. \[Stocks\] Earnings Surprise Analysis for Amazon (AMZN)

**Calculate the median 2-day percentage change in stock prices following
positive earnings surprises days.**

Steps: 
1. Load earnings data from CSV ([ha1_Amazon.csv](ha1_Amazon.csv))
containing earnings dates, EPS estimates, and actual EPS. Make sure you
are using the correct delimiter to read the data, such as in this
command `python pandas.read_csv("ha1_Amazon.csv", delimiter=';')` 
2. Download complete historical price data using yfinance 
3. Calculate 2-day percentage changes for all historical dates: for each sequence of 3 consecutive trading days (Day 1, Day 2, Day 3), compute the *return*as Close_Day3 / Close_Day1 - 1.(Assume Day 2 may correspond to the earnings announcement.) 
4. Identify positive earnings surprises (where “actual EPS \> estimated EPS” OR “Surprise (%)\>0”) 
5. Calculate 2-day percentage changes following positive earnings surprises. 
Show your answer in % (closest number to the 2nd digit): *return* \* 100.0 6.

(Optional) Compare the median 2-day percentage change for positive
surprises vs. all historical dates. Do you see the difference?

Context: Earnings announcements, especially when they exceed analyst
expectations, can significantly impact stock prices in the short term.

Reference: Yahoo Finance earnings calendar -
https://finance.yahoo.com/calendar/earnings?symbol=AMZN

*Additional*: Is there a correlation between the magnitude of the
earnings surprise and the stock price reaction? Does the market react
differently to earnings surprises during bull vs. bear markets?)


In [19]:
import pandas as pd
import numpy as np
import yfinance as yf

In [20]:
# Step 1: Load earnings data
amazon_earnings_df = pd.read_csv("../../ha1_Amazon.csv", sep=';')

# Preprocess: Clean and convert earnings date
def parse_date(d):
    try:
        return pd.to_datetime(d.split(" at")[0])
    except:
        return pd.NaT

amazon_earnings_df["Earnings Date"] = amazon_earnings_df["Earnings Date"].apply(parse_date)
amazon_earnings_df = amazon_earnings_df.dropna(subset=["Earnings Date"])
amazon_earnings_df.head()

Symbol           Company Earnings Date EPS Estimate Reported EPS  \
0   AMZN    Amazon.com Inc    2026-04-29            -            -   
1   AMZN    Amazon.com Inc    2026-02-04            -            -   
2   AMZN    Amazon.com Inc    2025-10-29            -            -   
3   AMZN    Amazon.com Inc    2025-07-30            -            -   
4   AMZN  Amazon.com, Inc.    2025-05-01       ???.36       ???.59   

  Surprise (%)  
0            -  
1            -  
2            -  
3            -  
4       +16.74

In [21]:
# Convert “Surprise (%)” strings (“-”, “+16.74”, etc.) to float, with “-” → NaN
def parse_surprise(val):
    if isinstance(val, str):
        val = val.strip()
        if val == "-" or val == "":
            return np.nan
        # Remove leading “+” if present, strip the “%” if present
        val = val.lstrip("+").rstrip("%").replace(",", ".")
        try:
            return float(val)
        except:
            return np.nan
    else:
        return np.nan
    
# Clean EPS Estimate and Reported EPS columns
def parse_eps(val):
    if isinstance(val, str):
        val = val.strip()
        val = val.replace("???", "0")
        try:
            return float(val)
        except:
            return np.nan
    return np.nan

    
amazon_earnings_df["Surprise (%)"] = amazon_earnings_df["Surprise (%)"].apply(parse_surprise)
amazon_earnings_df["EPS Estimate"] = amazon_earnings_df["EPS Estimate"].apply(parse_eps)
amazon_earnings_df["Reported EPS"] = amazon_earnings_df["Reported EPS"].apply(parse_eps)

positive_anonce_df = amazon_earnings_df[amazon_earnings_df["Surprise (%)"] > 0].copy().dropna()

In [22]:
amazon_df = yf.download("AMZN", start="2005-01-01", progress=False)[["Close"]]
amazon_df = amazon_df.rename_axis("Date").reset_index()

In [23]:
# Step 3: Compute 2-day percent changes for every trading date:
# Return = (Close on Day 3) / (Close on Day 1) – 1
amazon_df["Day1_Close"] = amazon_df["Close"]
amazon_df["Day3_Close"] = amazon_df["Close"].shift(-2)
amazon_df["2d_pct_change"] = (amazon_df["Day3_Close"] / amazon_df["Day1_Close"]) - 1

def find_next_trading_day(target_date, trading_series):
    future = trading_series[trading_series >= target_date]
    return future.min() if not future.empty else None

price_dates = amazon_df["Date"]
positive_changes = []

for _, row in positive_anonce_df.iterrows():
    earnings_date = row["Earnings Date"]
    day2 = find_next_trading_day(earnings_date, price_dates)
    if day2 is not None:
        idx = amazon_df.index[amazon_df["Date"] == day2]
        if len(idx) and idx[0] < len(amazon_df) - 2:  # ensure a 2-day window exists
            positive_changes.append(amazon_df.loc[idx[0], "2d_pct_change"])

historical_median = amazon_df["2d_pct_change"].median()
positive_median = np.nanmedian(positive_changes)

print(f"Median 2-day % change on ALL trading dates: {historical_median:.4%}")
print(f"Median 2-day % change AFTER positive earnings surprises: {positive_median:.4%}")

Median 2-day % change on ALL trading dates: 0.1806%
Median 2-day % change AFTER positive earnings surprises: -1.9855%


### Question 5. \[Exploratory, optional\] Brainstorm potential idea for your capstone project

**Free text answer**

Describe the capstone project you would like to pursue, considering your
aspirations, ML model predictions, and prior knowledge. Even if you are
unsure at this stage, try to generate an idea you would like to
explore-such as a specific asset class, country, industry vertical, or
investment strategy. Be as specific as possible.

*Example: I want to build a short-term prediction model for the
US/India/Brazil stock markets, focusing on the largest stocks over a
30-day investment horizon. I plan to use RSI and MACD technical
indicators and news coverage data to generate predictions.*


### Question 6. \[Exploratory, optional\] Investigate new metrics

**Free text answer**

Using the data sources we have covered (or any others you find
relevant), download and explore a few additional metrics or time series
that could be valuable for your project. Briefly explain why you think
each metric is useful. This does not need to be a comprehensive
list-focus on demonstrating your ability to generate data requests based
on your project description, identify and locate the necessary data, and
explain how you would retrieve it using Python.

## Submitting the solutions

Form for submitting:
https://courses.datatalks.club/sma-zoomcamp-2025/homework/hw01

## Leaderboard

Leaderboard link:
https://courses.datatalks.club/sma-zoomcamp-2025/leaderboard